firefly/kernel/intel64/cpu/cpu.cpp

```diff
@@ -3,22 +3,70 @@
 #include "firefly/intel64/cpu/ap/ap.hpp"
 #include "firefly/intel64/cpu/apic/apic.hpp"
 #include "firefly/intel64/int/interrupt.hpp"
+#include "firefly/memory-manager/allocator.hpp"
+#include "firefly/memory-manager/secondary/heap.hpp"
+#include "frg/manual_box.hpp"
+#include "frg/spinlock.hpp"
+#include "frg/vector.hpp"
 
-// Note: This is purposely bare-bones and lacking a
-// proper "per cpu" structure (i.e. gsbase).
-// There is no multi-core yet, but I'd prefer to structure
-// everything in a way so that it's easy to get it up and running now rather than later.
 namespace firefly::kernel {
-CpuData temporaryCpuInstance;
+namespace {
+frg::manual_box<CpuData> bootCpuData;
+frg::vector<CpuData *, Allocator> allCpuContexts;
+frg::ticket_spinlock allCpuContextsLock;
+}  // namespace
+
+void setupCpuContext(AssemblyCpuData *context) {
+    context->selfPointer = context;
+    wrmsr(MSR::GsBase, reinterpret_cast<uint64_t>(context));
+}
+
+void setupBootCpuContext() {
+    bootCpuData.initialize();
+    setupCpuContext(bootCpuData.get());
+}
+
+void initializeBootProccessor(uint64_t stack) {
+    setupBootCpuContext();
+    initializeThisCpu(stack);
+}
+
+void initializeApplicationProcessor(uint64_t stack) {
+    // TODO: Check with @vo1d how to do this
+    auto cpuData = new (mm::heap->allocate(sizeof(CpuData))) CpuData;
+    setupCpuContext(cpuData);
+    initializeThisCpu(stack);
+}
+
+CpuData *getCpuData(size_t k) {
+    return allCpuContexts[k];
+}
+
+CpuData *getCpuData() {
+    AssemblyCpuData *cpu_data;
+    asm volatile("mov %%gs:0, %0"
+                 : "=r"(cpu_data));
+    return reinterpret_cast<CpuData *>(cpu_data);
+}
 
 void initializeThisCpu(uint64_t stack) {
-    core::gdt::init(temporaryCpuInstance.gdt);
+    auto cpuData = getCpuData();
+
+    allCpuContextsLock.lock();
```

Comment:
Do we need locks here?
Does starting the AP's happen concurrently?

If so, we might want to lock `allCpuContexts.push(cpuData);`, perhaps even the entire function since we're setting up the TSS too.

Lmk what your thoughts are

Reply:
I had a concurrent startup in mind. Didn't get it to work correctly with console output though, boot CPU was keeping the lock.

Reply:
Can we remove the locks if they are unused then or no?

Reply:
I‘d let them there, we should rather figure out the serial stuff

Reply:
What's your reasoning for keeping locks in a synchronous function?

Also, what is the issue?
The console or serial logger?